src/opam_file.ml

```diff
@@ -87,6 +87,51 @@ module Create = struct
   let list f xs = List (nopos, List.map ~f xs)
   let string_list xs = list string xs
 
+  let normalise_field_order vars =
+    let normal_field_order = [
+      (* Extracted from opam/src/format/opamFile.ml *)
+      "opam-version";
+      "name";
+      "version";
+      "synopsis";
+      "description";
+      "maintainer";
+      "authors";
+      "author";
+      "license";
+      "tags";
+      "homepage";
+      "doc";
+      "bug-reports";
+      "depends";
+      "depopts";
+      "conflicts";
+      "conflict-class";
+      "available";
+      "flags";
+      "setenv";
+      "build";
+      "run-test";
+      "install";
+      "remove";
+      "substs";
+      "patches";
+      "build-env";
+      "features";
+      "messages";
+      "post-messages";
+      "depexts";
+      "libraries";
+      "syntax";
+      "dev-repo";
+      "pin-depends";
+      "extra-files" ]
+    in
+    List.filter_map normal_field_order
+      ~f:(fun field ->
+        Option.map (List.assoc vars field)
+          ~f:(fun v -> field, v))
```

Comment:
In addition to sorting the fields, this is also silently filtering out fields dune doesn't know about. That seems a bit dodgy given that we allow the user to add arbitrary fields via the template file. I suggest to put all unknown fields at the end in their original order.

Reply:
good catch! pushed a fix and some test cases for extension fields